library/posture.c

```diff
@@ -15,80 +15,123 @@ limitations under the License.
+
+struct pr_cb_ctx_s {
+    ziti_context ztx;
+    pr_info *info;
+};
+
+typedef struct pr_cb_ctx_s pr_cb_ctx;
 
-static void ziti_handle_mac(ziti_context ztx, char *id, char **mac_addresses, int num_mac);
 
-static void ziti_handle_domain(ziti_context ztx, char *id, char *domain);
+static void ziti_pr_ticker_cb(uv_timer_t *t);
 
-static void ziti_handle_os(ziti_context ztx, const char *id, char *os_type, char *os_version, char *os_build);
+static void ziti_pr_handle_mac(ziti_context ztx, char *id, char **mac_addresses, int num_mac);
 
-static void ziti_send_pr(ziti_context ztx);
+static void ziti_pr_handle_domain(ziti_context ztx, char *id, char *domain);
 
-static void ziti_send_pr_bulk(ziti_context ztx);
+static void ziti_pr_handle_os(ziti_context ztx, char *id, char *os_type, char *os_version, char *os_build);
 
-static void ziti_send_pr_individually(ziti_context ztx);
+static void ziti_pr_handle_process(ziti_context ztx, char *id, char *path, bool is_running, char *sha_512_hash, char **signers, int num_signers);
 
-static void
-ziti_handle_process(ziti_context ztx, const char *id, char *path, bool is_running, char *sha_512_hash, char **signers,
-                    int num_signers);
+static void ziti_pr_send(ziti_context ztx);
 
-extern void posture_init(struct ziti_ctx *ztx, long interval_secs) {
+static void ziti_pr_send_bulk(ziti_context ztx);
+
+static void ziti_pr_send_individually(ziti_context ztx);
+
+static bool ziti_pr_is_info_errored(ziti_context ztx, char *id);
+
+static void ziti_pr_free_pr_info_members(pr_info *info) {
+    FREE(info->id)
+    FREE(info->obj)
+}
+
+
+static void ziti_pr_free_pr_info(pr_info *info) {
+    ziti_pr_free_pr_info_members(info);
+    FREE(info)
+}
+
+static void ziti_pr_free_pr_cb_ctx(pr_cb_ctx *ctx) {
+    ziti_pr_free_pr_info(ctx->info);
+    FREE(ctx)
+}
+
+extern void ziti_posture_init(ziti_context ztx, long interval_secs) {
     if (ztx->posture_checks == NULL) {
         NEWP(pc, struct posture_checks);
 
         pc->interval = (double) interval_secs;
         pc->previous_session_id = NULL;
         pc->must_send_every_time = true;
         pc->must_send = false;
+
+        NEWP(error_states, struct model_map);
+        pc->error_states = error_states;
+
         ztx->posture_checks = pc;
     }
 
     if (!uv_is_active((uv_handle_t *) &ztx->posture_checks->timer)) {
         uv_timer_init(ztx->loop, &ztx->posture_checks->timer);
         ztx->posture_checks->timer.data = ztx;
-        uv_timer_start(&ztx->posture_checks->timer, ticker_cb, MILLIS(interval_secs), MILLIS(interval_secs));
+        uv_timer_start(&ztx->posture_checks->timer, ziti_pr_ticker_cb, MILLIS(interval_secs), MILLIS(interval_secs));
         uv_unref((uv_handle_t *) &ztx->posture_checks->timer);
     }
 }
 
 extern void ziti_posture_checks_free(struct posture_checks *pcs) {
-    model_map_clear(pcs->previous_responses, NULL);
-    model_map_clear(pcs->current_responses, NULL);
-    FREE(pcs);
-}
+    if (pcs != NULL) {
```

Comment:
it is not safe if pcs.timer is not closed

Reply:
Closed? Stopped? 

I was following the pattern I saw in `ziti.c` for `ziti_shutdown()` -> `ziti_ctx_free()`.

Where the `ziti_shutdown` I assumed would always be called first followed by a `ziti_ctx_free`.

Reply:
I can blindly call `uv_timer_stop` here though. It checks the state first to see if it needs stopping.

Reply:
updated

Reply:
stop is not enough, I believe, it's still referenced by the loop. you need to call `uv_close()`. freeing memory is not safe until close_cb is called (you can do it inside close_cb)

Reply:
I added `ziti_posture_checks_shutdown()` and call it before anything is freed (one line below this comment string).

Reply:
Fixed?